snapcraft/internal/remote_build/_launchpad.py

```diff
@@ -246,3 +264,87 @@ def _download_file(self, url: str, name: str) -> None:
         logger.debug("Download snap from {!r}".format(url))
         with urllib.request.urlopen(url) as response, open(name, "wb") as snapfile:
             shutil.copyfileobj(response, snapfile)  # type: ignore
+
+    def _finalize_source_tree_repository(self):
+        """Git-ify source tree."""
+        self._git_repo.index.add(["."])
+
+        timestamp = datetime.datetime.now().strftime("%Y-%m-%d %H:%M")
+        self._git_repo.index.commit(
+            "snapcraft commit\n\nversion: {}\ntimestamp: {}\n".format(
+                snapcraft.__version__, timestamp
+            )
+        )
+
+    def _init_source_tree_repository(self):
+        """Initialize empty git repo."""
+        if os.path.exists(self._repo_dir):
+            shutil.rmtree(self._repo_dir)
+
+        os.makedirs(self._repo_dir)
+        self._git_repo = git.Repo.init(self._repo_dir)
+
+    def _create_source_archive(self, source_dir: str, archive_path: str):
+        """Archive sources at source_dir into archive_path."""
+        logger.debug("creating source archive: {}".format(archive_path))
+        command = ["tar", "cJf", archive_path, "-C", source_dir, "."]
```

Comment:
so source will contain the original snap directory, which is desired, but how will it behave wrt git commands?

Reply:
Yeah, `snapcraft` itself appears to build fine (with its use of git-describe)

Reply:
My first spin of this was to extend the source classes to support mirroring.  It was a lot more optimal for build deltas, but I kept winding up with git-inside-of-git issues.

At first I just did the tarball hack for the obvious .git case, but reconsidered that any `local` sources may have other git-tracking-unfriendly files.  I don't currently know of any data loss issues with tarballing things up in this manner, but it's possible...